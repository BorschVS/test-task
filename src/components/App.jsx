import { lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Layout } from './Layout';

const Home = lazy(() => import('../pages/Home'));
const Flights = lazy(() => import('../pages/Flights'));
const Hotels = lazy(() => import('../pages/Hotels'));

const App = () => (
  <Routes>
    <Route path="/" element={<Layout />}>
      <Route index element={<Home />} />
      <Route path="flights" element={<Flights />} />
      <Route path="hotels" element={<Hotels />} />
    </Route>
  </Routes>
);

export default App;
