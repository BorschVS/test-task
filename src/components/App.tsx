import { FC, lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Layout } from 'components';

import { Page } from 'types/interfaces';

const Home = lazy(() => import('../pages/Home'));
const Flights = lazy(() => import('../pages/Flights'));
const Hotels = lazy(() => import('../pages/Hotels'));

const pages: Page[] = [
  { path: '/', element: <Home />, index: true },
  { path: 'flights', element: <Flights /> },
  { path: 'hotels', element: <Hotels /> },
];

const App: FC = () => (
  <Routes>
    <Route path="/" element={<Layout />}>
      {pages.map((page) =>
        page.index ? (
          <Route key={page.path} index element={page.element} />
        ) : (
          <Route key={page.path} path={page.path} element={page.element} />
        )
      )}
    </Route>
  </Routes>
);

export default App;
