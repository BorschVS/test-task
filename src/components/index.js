export { default as App } from './App';
export { default as CategoryFilter } from './CategoryFilter';
export { default as FlightCard } from './FlightCard';
export { default as TransferFilter } from './TransferFilter';
export { default as ResponsiveAppBar } from './ResponsiveAppBar';
