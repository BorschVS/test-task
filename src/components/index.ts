export { default as App } from './App';
export { default as Layout } from './Layout';
export { default as CategoryFilter } from './CategoryFilter';
export { default as FlightCard } from './FlightCard';
export { default as TransferFilter } from './TransferFilter';
export { default as ResponsiveAppBar } from './ResponsiveAppBar';
export { default as ContactModal } from './ContactModal';
export { default as Form } from './Form';
export { default as AvatarImage } from './AvatarImage';
export { default as BackgroundImage } from './BackgroundImage';
