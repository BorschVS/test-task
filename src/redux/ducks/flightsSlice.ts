import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { FlightData, FlightsState } from 'types/interfaces';
import { FlightCategory } from 'types/types';

import { FAST_VALUE, CHEAP_VALUE } from '../../constants';

export const GET_FLIGHTS = 'GET_FLIGHTS';
export const SET_FLIGHTS = 'SET_FLIGHTS';
export const SET_FLIGHTS_LOADING = 'SET_FLIGHTS_LOADING';
export const SET_FLIGHTS_ERROR = 'SET_FLIGHTS_ERROR';
export const SET_STOPS_FILTER = 'SET_STOPS_FILTER';
export const SET_ALL_STOPS = 'SET_ALL_STOPS';
export const SET_STOPS_FILTER_STATUS = 'SET_STOPS_FILTER_STATUS';
export const SET_CATEGORY_FILTER = 'SET_CATEGORY_FILTER';

const initialState: FlightsState = {
  availableFlights: [],
  filteredFlights: [],
  stopsFilter: [],
  stopsFilterStatus: false,
  flightCategoryFilter: CHEAP_VALUE,
  loading: false,
  error: null,
};

const flightsSlice = createSlice({
  name: 'flights',
  initialState,
  reducers: {
    setFlights(state, action: PayloadAction<FlightData[]>) {
      state.availableFlights = action.payload
        .slice()
        .sort((prev, next) => prev.price - next.price);
    },
    setFlightsLoading(state, action: PayloadAction<boolean>) {
      state.loading = action.payload;
    },
    setFlightsError(state, action: PayloadAction<Error | string>) {
      state.error = action.payload;
      state.loading = false;
    },
    setCategoryFilter(state, action: PayloadAction<FlightCategory>) {
      if (action.payload === CHEAP_VALUE) {
        state.flightCategoryFilter = action.payload;
        state.filteredFlights = state.filteredFlights
          .slice()
          .sort((prev, next) => prev.price - next.price);
        state.availableFlights = state.availableFlights
          .slice()
          .sort((prev, next) => prev.price - next.price);
      }
      if (action.payload === FAST_VALUE) {
        state.flightCategoryFilter = action.payload;
        state.filteredFlights = state.filteredFlights
          .slice()
          .sort(
            (prev, next) =>
              prev.segments[0].duration - next.segments[0].duration
          );
        state.availableFlights = state.availableFlights
          .slice()
          .sort(
            (prev, next) =>
              prev.segments[0].duration - next.segments[0].duration
          );
      }
    },
    setStopsFilter(state, action: PayloadAction<number[]>) {
      state.stopsFilter = action.payload;
      state.filteredFlights = state.availableFlights.filter((flight) =>
        flight.segments.some((segment) =>
          action.payload.includes(segment.stops.length)
        )
      );
      state.stopsFilterStatus = false;
    },
    setAllStops(state) {
      state.stopsFilter = [0, 1, 2, 3];
      state.filteredFlights = state.availableFlights;
      state.stopsFilterStatus = true;
    },
    setStopsFilterStatus(state) {
      state.stopsFilterStatus = state.stopsFilter.length > 0;
    },
  },
});

export const {
  setFlights,
  setFlightsLoading,
  setFlightsError,
  setCategoryFilter,
  setStopsFilter,
  setAllStops,
  setStopsFilterStatus,
} = flightsSlice.actions;

export default flightsSlice.reducer;
