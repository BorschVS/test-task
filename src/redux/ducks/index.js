export { default as flights } from './flights';
export { default as form } from './form';
export { default as modal } from './modal';
export { default as images } from './images';