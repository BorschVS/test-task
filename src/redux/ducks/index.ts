export { flightsReducer } from './flights';
export { formReducer } from './form';
export { modalReducer } from './modal';
export { imagesReducer } from './images';
