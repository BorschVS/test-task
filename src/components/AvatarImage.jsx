import { useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { Avatar } from '@mui/material';

import { getImageRequest } from '../redux/ducks/images';

const AvatarImage = ({
  imageName,
  alt,
  component = 'div',
  width = 40,
  height = 40,
  margin = '0',
}) => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getImageRequest(imageName));
  }, [dispatch, imageName]);

  return (
    <span>
      <Avatar
        component={component}
        src={`/images/${imageName}`}
        alt={alt}
        sx={{
          display: 'block',
          margin,
          width,
          height,
        }}
      />
    </span>
  );
};

export default AvatarImage;
