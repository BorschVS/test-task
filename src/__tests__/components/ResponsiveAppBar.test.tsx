import { screen, fireEvent, render, cleanup } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { ResponsiveAppBar } from 'components';

import { GET_FLIGHTS } from '../../redux/ducks/flightsSlice';

jest.mock('react-redux', () => ({
    useDispatch: jest.fn(),
}));

const renderWithRouter = () => render(<MemoryRouter initialEntries={['/']}><ResponsiveAppBar /></MemoryRouter>);

describe('ResponsiveAppBar component', () => {
    const mockDispatch = jest.fn();

    beforeEach(() => {
        (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    });

    afterEach(() => {
        cleanup();
        jest.clearAllMocks();
    });

    test('Should render ResponsiveAppBar correctly', () => {
        renderWithRouter();

        const appBar = screen.getByTestId('responsive-app-bar')
        expect(appBar).toBeInTheDocument();
    });

    test('Should dispatch GET_FLIGHTS action on "Flights" button click', () => {
        renderWithRouter();

        const flightsButtons = screen.getAllByTestId('flights-link');
        flightsButtons.forEach(button => fireEvent.click(button));

        expect(mockDispatch).toHaveBeenCalledWith({ type: GET_FLIGHTS });
    });

    test('Should render "Flights" link', () => {
        renderWithRouter();

        const links = screen.getAllByTestId('flights-link');
        links.forEach(link => expect(link).toBeInTheDocument())
    });

    test('Should render "Hotels" link', () => {
        renderWithRouter();

        const links = screen.getAllByTestId('hotels-link');
        links.forEach(link => expect(link).toBeInTheDocument());
    });

    test('Should open mobile menu and expect Flights links', () => {
        renderWithRouter();

        const menuButton = screen.getByLabelText('account of current user');
        fireEvent.click(menuButton);

        const links = screen.getAllByText('Flights');
        links.forEach(link => expect(link).toBeInTheDocument());
    });

    test('Should open mobile menu and expect Hotels links', () => {
        renderWithRouter();

        const menuButton = screen.getByLabelText('account of current user');
        fireEvent.click(menuButton);

        const links = screen.getAllByText('Hotels');
        links.forEach(link => expect(link).toBeInTheDocument());
    });

    test('Should match ResponsiveAppBar snapshot', () => {
        const { asFragment } = renderWithRouter();

        expect(asFragment()).toMatchSnapshot();
    });
});
