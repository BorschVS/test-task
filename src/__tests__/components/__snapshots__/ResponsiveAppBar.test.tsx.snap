// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsiveAppBar component Should match ResponsiveAppBar snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorInfo MuiAppBar-positionStatic css-1wwtsp3-MuiPaper-root-MuiAppBar-root"
    data-testid="responsive-app-bar"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
      >
        <a
          data-testid="home-link"
          href="/"
        >
          <span>
            <div
              class="MuiAvatar-root MuiAvatar-circular css-gh7mkj-MuiAvatar-root"
            >
              <img
                alt="Lets fly logo"
                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="/images/plane.png"
              />
            </div>
          </span>
        </a>
        <div
          class="MuiBox-root css-15ytkbm"
        >
          <a
            data-testid="flights-link"
            href="/flights"
            style="text-decoration: none;"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1q39md6-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Flights
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
          <a
            data-testid="hotels-link"
            href="/hotels"
            style="text-decoration: none;"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1q39md6-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Hotels
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
        </div>
        <div
          class="MuiBox-root css-iqpydz"
        >
          <button
            aria-controls="menu-appbar"
            aria-haspopup="true"
            aria-label="account of current user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MenuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
